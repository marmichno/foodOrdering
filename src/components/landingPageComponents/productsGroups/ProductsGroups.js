import {motion, AnimatePresence} from 'framer-motion';
import {IoMdArrowDropdownCircle} from 'react-icons/io';
import {useState, useEffect} from 'react';
import {useDispatch, useSelector} from 'react-redux';
import {groupGetRequest} from '../../../requests/groupGetRequest';
import {changeOrderGroupId} from '../../../actions';

const containerVariants = {
    hidden: {
        x: -2000
    },
    visible: {
        x: 0,
        transition: {
            type: 'spring',
            delay: 0.5,
            mass: 1,
            damping: 12
        }
    },
    exit: {
        x:'-100vw',
        transition: {ease: 'easeInOut', duration:0.5}
    }
}

const dropdownVariants = {
    hidden: {
        y: -500
    },
    visible: {
        y: 0,
        transition: {
            type: 'spring',
            mass: 1,
            damping: 15
        }
    },
    exit: {
        scale:0.01,
        transition: {ease: 'easeInOut', duration:0.5}
    }
}

export const ProductsGroups = () => {

    const productGroupId = useSelector(state => state.orderGroupIdReducer);
    const dispatch = useDispatch();

    const [selectedGroup, setSelectedGroup] = useState([]);
    const [productGroups, setProductGroups] = useState([]);
    const [showProductGroups, setShowProductGroups] = useState(false);
    const [groupId, setGroupId] = useState(productGroupId);

    useEffect(() =>{
        getGroups();
    }, []);

    useEffect(() => {
        dispatch(changeOrderGroupId(groupId));
    },[groupId]);

    const getGroups = async () =>{
        const response = await groupGetRequest();
        if(response !== undefined){
            setProductGroups(response);
            setSelectedGroup(response[0].name);
            setGroupId(response[0].id);
        }else{
            return;
        }
    }

    const changeActive = (e) => {
        const allHeaders = document.querySelectorAll('.foodHeadersContainer .foodTypesContainer .dropdown .toggle');
        allHeaders.forEach(element => element.classList.remove('active'));
        e.target.classList.add('active');
        setSelectedGroup(e.currentTarget.innerHTML);
        setGroupId(e.currentTarget.dataset.id);
    }

    return(
            <motion.div
            variants={containerVariants}
            initial="hidden"
            animate="visible"
            exit="exit"
            className="foodHeadersContainer">
                <div className="foodGroupsContainer">
                    <div className="list">
                        <div className="selected"><h1 data-testid="selected">{selectedGroup !== undefined ? selectedGroup : null}</h1><hr></hr></div>
                        <div className="toggle" data-testid="dropDown" onClick={() => setShowProductGroups(!showProductGroups)}><IoMdArrowDropdownCircle/></div>
                        <AnimatePresence>
                            {showProductGroups === true ?
                                <motion.div className="dropdown"
                                    variants={dropdownVariants}
                                    initial="hidden"
                                    animate="visible"
                                    exit="exit"
                                    >
                                    {productGroups !== undefined ?
                                    productGroups.map((value, index) => {
                                        return <div key={value.id} className="dropdownList"><h1 data-id={value.id} data-testid={`group${index}`} onClick={changeActive}>{value.name}</h1></div>
                                    })
                                    : null
                                    }
                                </motion.div>
                            : null
                        }
                        </AnimatePresence>
                    </div>
                </div>
            </motion.div>
        )
}

export default ProductsGroups;