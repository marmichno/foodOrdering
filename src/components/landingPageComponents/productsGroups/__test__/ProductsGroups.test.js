import ProductsGroups from '../../productsGroups/ProductsGroups';
import validProductGroupJson from '../__mocks__/validProductGroup.json';
import mockedAxios from 'axios';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux'; 
import {render, fireEvent, waitFor} from '@testing-library/react';
import TestRenderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import React from "react";


const store = createStore(() => [], {}, applyMiddleware());

test("checks if product group is visible after api get call", async () => {
    mockedAxios.get.mockResolvedValueOnce({data: validProductGroupJson});

    const {getByTestId, getByText} = render(<Provider store={store}><ProductsGroups/></Provider>);

    await waitFor(() => {
        expect(getByText('sushi rolls'));
    })
});

test("check if other product groups than default are showing after onClick event", async() => {
    mockedAxios.get.mockResolvedValueOnce({data: validProductGroupJson});

    const {getByTestId, getByText} = render(<Provider store={store}><ProductsGroups/></Provider>);

    fireEvent.click(getByTestId('dropDown'))

    await waitFor(() =>{
        expect(getByText('prawns'));
    })
});

test("check if produdct group is changed after click on other group from dropdown", async() => {
    mockedAxios.get.mockResolvedValueOnce({data: validProductGroupJson});

    const {getByTestId} = render(<Provider store={store}><ProductsGroups/></Provider>);

    fireEvent.click(getByTestId('dropDown'));

    await waitFor(() =>{
        fireEvent.click(getByTestId('group1'));
    });

    await waitFor(() =>{
        expect(getByTestId('selected').textContent).toBe("prawns");
    })
});

test("Initial render should call dispatch one time after api call", async() => {
    mockedAxios.get.mockResolvedValueOnce({data: validProductGroupJson});

    const {getByTestId} = render(<Provider store={store}><ProductsGroups/></Provider>);

    store.dispatch = jest.fn();

    await waitFor(() =>{
        expect(store.dispatch).toHaveBeenCalledTimes(1);
    })
});

test("On product group click dispatch should be called", async() => {
    mockedAxios.get.mockResolvedValueOnce({data: validProductGroupJson});

    const {getByTestId} = render(<Provider store={store}><ProductsGroups/></Provider>);

    store.dispatch = jest.fn();

    fireEvent.click(getByTestId('dropDown'));

    await waitFor(() =>{
        fireEvent.click(getByTestId('group1'));
    });

    await waitFor(() =>{
        expect(store.dispatch).toHaveBeenCalledTimes(2);
    })
});

test("matches snapshot", () => {
    const tree = TestRenderer.create(<Provider store={store}><ProductsGroups/></Provider>).toJSON();
    expect(tree).toMatchSnapshot();
});