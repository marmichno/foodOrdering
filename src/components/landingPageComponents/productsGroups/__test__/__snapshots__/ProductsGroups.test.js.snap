// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="foodHeadersContainer"
  style={
    Object {
      "transform": "translateX(-2000px) translateZ(0)",
    }
  }
>
  <div
    className="foodGroupsContainer"
  >
    <div
      className="list"
    >
      <div
        className="selected"
      >
        <h1
          data-testid="selected"
        />
        <hr />
      </div>
      <div
        className="toggle"
        data-testid="dropDown"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 48C141.125 48 48 141.125 48 256s93.125 208 208 208 208-93.125 208-208S370.875 48 256 48zm0 272l-96-96h192l-96 96z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
